pkg/agent/nplagent/portcache/port_table.go

```diff
@@ -0,0 +1,180 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portcache
+
+import (
+	"fmt"
+	"sync"
+
+	nplutils "github.com/vmware-tanzu/antrea/pkg/agent/nplagent/lib"
+	"github.com/vmware-tanzu/antrea/pkg/agent/nplagent/rules"
+	"k8s.io/klog"
+)
+
+const (
+	successStatus = 0
+	pendingStatus = 1
+	failStatus    = 2
+)
+
+type NodePortData struct {
+	Nodeport int
+	Podport  int
+	Podip    string
+	Status   int
+	//podname  string
+}
+
+type PortTable struct {
+	Table        map[int]NodePortData
+	StartPort    int
+	EndPort      int
+	PodPortRules rules.PodPortRules
+	tableLock    sync.RWMutex
+}
+
+var once sync.Once
+var ptable PortTable
+
+func NewPortTable(start, end int) *PortTable {
+	once.Do(func() {
+		ptable = PortTable{StartPort: start, EndPort: end}
+		ptable.Table = make(map[int]NodePortData)
+		ptable.PodPortRules = rules.Initrules()
+	})
+	return &ptable
+}
+
+func GetPortTable() *PortTable {
+	return &ptable
+}
+
+func (pt *PortTable) PopulatePortTable(r rules.PodPortRules) {
+	portMap := make(map[int]string)
+	ok, err := r.GetAllRules(portMap)
+	if !ok {
+		klog.Warningf("Could not populate port table cache due to error: %v", err)
+		return
+	}
+	table := make(map[int]NodePortData)
+	for nodeport, podip := range portMap {
+		entry := NodePortData{
+			Nodeport: nodeport,
+			Podip:    podip,
+		}
+		table[nodeport] = entry
+	}
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	pt.Table = table
+}
+
+func (pt *PortTable) AddUpdateEntry(nodeport, podport int, podip string) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	data := NodePortData{Nodeport: nodeport, Podport: podport, Podip: podip}
+	pt.Table[nodeport] = data
+}
+
+func (pt *PortTable) DeleteEntry(nodeport int) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	delete(pt.Table, nodeport)
+}
+
+func (pt *PortTable) DeleteEntryByPodIP(ip string) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	for i, data := range pt.Table {
+		if data.Podip == ip {
+			delete(pt.Table, i)
+		}
+	}
+}
+
+func (pt *PortTable) DeleteEntryByPodIPPort(ip string, port int) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	for i, data := range pt.Table {
+		if data.Podip == ip && data.Podport == port {
+			delete(pt.Table, i)
+		}
+	}
+}
+
+func (pt *PortTable) GetEntry(nodeport int) *NodePortData {
+	pt.tableLock.RLock()
+	defer pt.tableLock.RUnlock()
+	data, _ := pt.Table[nodeport]
+	return &data
+}
+
+func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+	pt.tableLock.RLock()
+	defer pt.tableLock.RUnlock()
+	for _, data := range pt.Table {
+		if data.Podip == ip && data.Podport == port {
+			return &data
+		}
+	}
+	return nil
+}
+
+func (pt *PortTable) getFreePort() int {
+	pt.tableLock.RLock()
+	defer pt.tableLock.RUnlock()
+	for i := pt.StartPort; i <= pt.EndPort; i++ {
+		if _, ok := pt.Table[i]; !ok && nplutils.IsPortAvailable(i) {
```

Comment:
I'm trying to understand the usefulness of the `IsPortAvailable` port here. Once we select a port number for NPL and install the iptables NAT rule, is there anything preventing a process running on the host to bind to this port as well?